pol-core/pol/dap/server.cpp

```diff
@@ -48,34 +76,184 @@ class SocketReaderWriter : public dap::ReaderWriter
+            return dap::Error( "No script attached." );
+
+          UOExecutor* exec = _uoexec_wptr.get_weakptr();
+          dap::ThreadsResponse response;
+          dap::Thread thread;
+          thread.id = 1;
+          thread.name = Clib::tostring( exec->pid() );
+          response.threads.push_back( thread );
+          return response;
+        } );
+
+    _session->registerHandler( [&]( const dap::SetExceptionBreakpointsRequest& )
+                               { return dap::SetExceptionBreakpointsResponse(); } );
+
+    _session->registerHandler(
+        [&]( const dap::ContinueRequest& ) -> dap::ResponseOrError<dap::ContinueResponse>
+        {
+          if ( !_uoexec_wptr.exists() )
+          {
+            return dap::Error( "No script attached." );
+          }
+
+          UOExecutor* uoexec = _uoexec_wptr.get_weakptr();
+          if ( !uoexec->in_debugger_holdlist() )
+          {
+            return dap::Error( "Script not ready to trace." );
+          }
+
+          uoexec->dbg_run();
+          uoexec->revive_debugged();
+          return dap::ContinueResponse{};
+        } );
+
+    _session->registerHandler(
+        [&]( const dap::StackTraceRequest& ) -> dap::ResponseOrError<dap::StackTraceResponse>
+        { return dap::StackTraceResponse{}; } );
+
+    _session->registerHandler(
+        [&]( const dap::ScopesRequest& ) -> dap::ResponseOrError<dap::ScopesResponse>
+        { return dap::ScopesResponse{}; } );
+
+    _session->registerHandler(
+        [&]( const dap::VariablesRequest& ) -> dap::ResponseOrError<dap::VariablesResponse>
+        { return dap::VariablesResponse{}; } );
+  };
+
+  // Session event handlers added before initialization
   _session->registerHandler(
-      []( const dap::InitializeRequest& )
+      [&]( const dap::PolInitializeRequest& request )
+          -> dap::ResponseOrError<dap::InitializeResponse>
       {
-        POLLOG_INFO << "Got InitializeRequest.\n";
+        if ( !Plib::systemstate.config.debug_password.empty() &&
+             request.password.value( "" ) != Plib::systemstate.config.debug_password )
+          return dap::Error( "Password not recognized." );
+
+        attach_authorized_handlers();
         return dap::InitializeResponse{};
       } );
 
+  _session->registerHandler(
+      [&]( const dap::DisconnectRequest& )
+      {
+        if ( _uoexec_wptr.exists() )
+        {
+          UOExecutor* exec = _uoexec_wptr.get_weakptr();
+          exec->detach_debugger();
+        }
+
+        return dap::DisconnectResponse();
+      } );
+
+  _session->registerSentHandler(
+      [&]( const dap::ResponseOrError<dap::InitializeResponse>& response )
+      {
+        if ( response.error )
+          _rw->close();
+        else
+          _session->send( dap::InitializedEvent() );
+      } );
+
+  // Error handler
+  _session->onError(
+      [&]( const char* msg )
+      {
+        POLLOG_ERROR << "Debugger session error: " << msg << "\n";
+        _rw->close();
```

Comment:
Wrong spot, but I need a changed line:
The SocketReaderWriter is used in 2 threads the dap thread and the main thread for closing. Haven't checked but I assume that the socket does not support multi threading. I would suggest to have in each method a eg SpinLock to prevent that close gets called while the other threads performs a read/write